### ISSUE TITLE
Performance regression (requests per second) from 0.6.20 to 0.7.2 when using many routes

### ISSUE THREAD

**Issue by kapteinstein** [tokio-rs/axum#2449]
- [x] I have looked for existing issues (including closed) about this

### Version

```
├── axum v0.7.2
│   ├── axum-core v0.4.1
```

### Platform

```
Darwin 23.2.0 Darwin Kernel Version 23.2.0: Wed Nov 15 21:54:10 PST 2023; root:xnu-10002.61.3~2/RELEASE_X86_64 x86_64
```

### Description

The throughput (requests per second) seems to be reduced if I have many routes specified after upgrading to axum 0.7.2. This is not an issue with axum 0.6.20. I have made a small example/pseudo code to reproduce the unintended behaviour:

```
use axum::{
    routing::get,
    Router,
};

async fn hello() -> String {
    String::from("hello")
}

#[tokio::main]
async fn main() {
    // build our application with a single route
    let app = Router::new()
        .route("/", get(|| async { "Hello, World!" }))
        .route("/0", get(hello))
        .route("/1", get(hello))
        .route("/2", get(hello))
        .route("/3", get(hello))
        .route("/4", get(hello))
        .route("/5", get(hello))
        .route("/6", get(hello))
        // ... skipped the middle lines here on GitHub, 
        // but the pattern is followed to reproduce the unintended behaviour (1000 routes)
        .route("/996", get(hello))
        .route("/997", get(hello))
        .route("/998", get(hello))
        .route("/999", get(hello));

    // I used this for axum 0.7.2
    let listener = tokio::net::TcpListener::bind("[::1]:3444").await.unwrap();
    axum::serve(listener, app).await.unwrap();

    // I used this for axum 0.6.20
    axum::Server::bind(&"[::1]:3444".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
```

I used `oha` to measure the throughput: `oha --ipv6 --no-tui -z 10s http://localhost:3444` 

for axum 0.7.2:
```
  Success rate:	1.0000
  Total:	10.1465 secs
  Slowest:	0.0312 secs
  Fastest:	0.0006 secs
  Average:	0.0030 secs
  Requests/sec:	13334.5141
```

for axum 0.6.20:
```
  Success rate:	1.0000
  Total:	10.1471 secs
  Slowest:	0.0302 secs
  Fastest:	0.0001 secs
  Average:	0.0009 secs
  Requests/sec:	48525.3179
```

Using fewer routes the requests per second is similar to axum 0.6.20. 

**EDIT:** I did a git bisect and the problem seems to start occurring at around commit b34715fe81bea0a3650782c38e9ae8922d74f73c (I wasn’t able to test all commits)

**Comment by davidpdrsn** [comment]
The fact that it started happening around https://github.com/tokio-rs/axum/pull/2094 is very strange. That PR doesn't really do much that could impact performance. I'm wondering if the problem could perhaps be in hyper or hyper-util.

Would be curious to know if it can be reproduced without using axum and instead comparing hyper 0.14 and hyper 1.0 + hyper-util.

**Comment by kapteinstein** [comment]
Please take that commit id with a grain of salt. I wasn’t able to test all commits (had to skip quite a few because of compilation errors) so I might have skipped the “real” commit that introduced the issue.

**Comment by fundon** [comment]
https://github.com/tokio-rs/axum/blob/5d388c8d91440bc2865a894a3587971a4e7bc47c/axum/src/serve.rs#L455

Feels heavy cloning in `tokio::spawn`.

**Comment by davidpdrsn** [comment]
@fundon did you benchmark it?

**Comment by fundon** [comment]
I'm trying to write a test, but it's hard. `oneshot` needs `self`.

**Comment by davidpdrsn** [comment]
The reason for that clone is that hyper's `Service` trait uses `&self` whereas tower uses `&mut self`. So we have to clone the router to get a `&mut Router` that we can `call`.

**Comment by davidpdrsn** [comment]
@kapteinstein would you be up for testing https://github.com/tokio-rs/axum/pull/2476? That makes `Router` cheaper to `Clone` so I'm curious if that helps with the regression you're seeing.

**Comment by kapteinstein** [comment]
Yes, that seems to have solved the issue! Some benchmarks on my computer testing out the different commits:

tag `axum-v0.6.20`:
```
  Success rate:	1.0000
  Total:	10.1475 secs
  Slowest:	0.0295 secs
  Fastest:	0.0001 secs
  Average:	0.0009 secs
  Requests/sec:	47711.1376
```

tag `axum-v0.7.2`:
```
  Success rate:	1.0000
  Total:	10.1461 secs
  Slowest:	0.0769 secs
  Fastest:	0.0005 secs
  Average:	0.0029 secs
  Requests/sec:	14707.6633
```

branch `david/arc-router`:
```
  Success rate:	1.0000
  Total:	10.1425 secs
  Slowest:	0.0050 secs
  Fastest:	0.0002 secs
  Average:	0.0009 secs
  Requests/sec:	47294.4575
```

**Comment by davidpdrsn** [comment]
Awesome! Then I feel pretty confident in saying the regression was caused by the additional cloning of the router. Like I mentioned in https://github.com/tokio-rs/axum/issues/2449#issuecomment-1872532236 there is currently no way around that.

I think I'll close this issue for now. We wanna ship 0.8 relatively soon-ish (end of Q1 2024) to take advantage of AFIT so this will be solved together with that.

**Comment by davidpdrsn** [comment]
The fix for this was shipped in axum 0.7.4 🎉
